### PR TITLE
snap-exec: add proper integration test for snap-exec

### PR DESCRIPTION
**Pull request by mvo5**
It also fixes the error messages and fixes another off-by-one error that got overlooked in the previous fix.

### CODE REVIEW

File: cmd/snap-exec/main.go [hunk 1]
```diff
@@ -96,15 +96,15 @@ func findCommand(app *snap.AppInfo, command string) (string, error) {
...
+		return fmt.Errorf("cannot parse revision %q: %s", revision, err)
 	}
 
 	snapName, appName := snap.SplitSnapApp(snapApp)
 	info, err := snap.ReadInfo(snapName, &snap.SideInfo{
 		Revision: rev,
 	})
 	if err != nil {
-		return err
+		return fmt.Errorf("cannot read info for %q", snapName, err)
```
Comment (pedronis) [inline cmd/snap-exec/main.go]: missing ": %s" or ": %v"

File: cmd/snap-exec/main.go [hunk 2]
```diff
@@ -114,7 +114,7 @@ func snapExec(snapApp, revision, command string, args []string) error {
 
 	cmd, err := findCommand(app, command)
 	if err != nil {
-		return err
+		return fmt.Errorf("cannot find command %q in app %q: %s", command, app, err)
```
Comment (pedronis) [inline cmd/snap-exec/main.go]: it will read a bit funny when command is "" but not sure it's important

File: cmd/snap-exec/main_test.go [hunk 3]
```diff
@@ -134,3 +142,52 @@ func (s *snapExecSuite) TestSnapExecErrorsOnMissingSnapApp(c *C) {
...
+	syscallExec = func(argv0 string, argv []string, env []string) error {
+		cmd := exec.Command(argv[0], argv[1:]...)
+		cmd.Env = env
+		cmd.Stdout = os.Stdout
+		return cmd.Run()
+	}
+
+	// run it
+	os.Args = []string{"snap-exec", "snapname.app", "foo", "--bar=baz", "foobar"}
+	err = run()
```
Comment (pedronis) [inline cmd/snap-exec/main_test.go]: thanks for this kind of test!

File: cmd/snap-exec/main_test.go [hunk 4]
```diff
@@ -134,3 +142,52 @@ func (s *snapExecSuite) TestSnapExecErrorsOnMissingSnapApp(c *C) {
...
+		"done\n"+
+		"printf \"\\n\" >> %[1]q\n", canaryFile)), 0755)
+	c.Assert(err, IsNil)
+
+	// we can not use the real syscall.execv here because it would
+	// replace the entire test :)
+	syscallExec = func(argv0 string, argv []string, env []string) error {
+		cmd := exec.Command(argv[0], argv[1:]...)
+		cmd.Env = env
+		cmd.Stdout = os.Stdout
```
Comment (pedronis) [inline cmd/snap-exec/main_test.go]: shouldn't we use a buffer here and sanity check that is empty after?

Comment (mvo5) [inline cmd/snap-exec/main_test.go]: Yes, excellent point.

File: cmd/snap-exec/main.go (hunk 2 above)
Comment (pedronis) [inline cmd/snap-exec/main.go]: actually we control the findCommand errors in the same file,  so return err would be ok

we might tweak them though if we want

**Comment by chipaca** [comment]
:+1:

**Pull request merged**